pennylane/transforms/optimization/pattern_matching.py

```diff
@@ -0,0 +1,1594 @@
+
+                        pred = group.pred_block
+
+                        # Choose the first configuration
+                        qubit = group.qubit_config[0]
+
+                        # First add all the predecessors of the given match.
+                        for elem in pred:
+                            node = circuit_dag.get_node(elem)
+                            inst = copy.deepcopy(node.op)
+                            apply(inst)
+                            already_sub.append(elem)
+
+                        already_sub = already_sub + circuit_sub
+
+                        # Then add the inverse of the template.
+                        for index in template_inverse:
+                            all_qubits = tape.wires.tolist()
+                            all_qubits.sort()
+                            wires_t = group.template_dag.get_node(index).wires
+                            wires_c = [qubit[x] for x in wires_t]
+                            wires = [all_qubits[x] for x in wires_c]
+
+                            node = group.template_dag.get_node(index)
+                            inst = copy.deepcopy(node.op)
+                            inst._wires = Wires(wires)
+                            inst.adjoint()
+
+                    # Add the unmatched gates.
+                    for node_id in substitution.unmatched_list:
+                        node = circuit_dag.get_node(node_id)
+                        inst = copy.deepcopy(node.op)
+                        apply(inst)
+
+            tape = tape_inside
+
+    for op in tape.operations:
+        op._wires = Wires([inverse_wires_map[wire] for wire in op.wires.tolist()])
+        apply(op)
+
+    # After optimization, simply apply the measurements
+    for obs in observables:
+        obs._wires = Wires([inverse_wires_map[wire] for wire in obs.wires.tolist()])
+
+    for m in measurements:
+        apply(m)
+
+
+def pattern_matching(circuit_dag, pattern_dag):
+    r"""Function that applies the pattern matching algorithm and returns the list of maximal matches.
+
+    Args:
+        circuit_dag (.CommutationDAG): A commutation DAG representing the circuit to be optimized.
+        pattern_dag(.CommutationDAG): A commutation DAG representing the pattern.
+
+    Returns:
+        list(Match): the list of maximal matches.
+    """
+    # Match list
+    match_list = []
+
+    # Loop through all possible initial matches
+    for node_c in circuit_dag.get_nodes():
+        for node_p in pattern_dag.get_nodes():
+            # Initial matches between two identical gates (No qubits comparison)
+            if _compare_operation_without_qubits(node_c[1], node_p[1]):
+
+                circuit_range = range(0, circuit_dag.num_wires)
+
+                # Fix qubits from the first (target fixed and control restrained)
+                not_fixed_qubits_confs = _not_fixed_qubits(
+                    circuit_range, node_c[1].wires, pattern_dag.num_wires - len(node_p[1].wires)
```

Comment:
Instead of having to define `circuit_range` above that only gets used here, maybe we could pass in `circuit_dag.num_wires` instead.  

This may also be a situation where pulling this code out to a helper function obscures more than it clarifies.

I generally think that unless an utility function is used many times (it's only used once here) or the name of the function makes a large difference in readability (I still had to track down the function to figure out what it was doing), one or two line functions are better avoided.  They increase obscurity and complexity without much improvement in brevity.